service/dataset.py

```diff
@@ -48,60 +76,151 @@ def read(dataset_request):
+        if dataset_request["dataset_name"] in OWNERS[client_guid]:
+            return dataset
+
+    # Checks to see if authorized users specified
+    if 'authorized_users' in dataset and len(dataset['authorized_users']) > 0:
+        if client_guid not in dataset['authorized_users']:
+            abort(404, "User not authorized to access this dataset.")
+
+    # Track readrelease
+    USERS_RELEASED[dataset_name][client_guid] += 1
+
+    # Decrement budget if this is a users first read
+    if USERS_RELEASED[dataset_name][client_guid] > 1:
+        adjusted_budget = dataset["budget"] - dataset_request["budget"]
+        if adjusted_budget >= 0.0:
+            dataset["budget"] = adjusted_budget
+        else:
+            abort(412, "Not enough budget for read. Remaining budget: {}".format(dataset["budget"]))
+    
     return dataset
 
+def release(release_request):
+    """Releases a new dataset
+
+    :param release_request: The dataset to read and budget to use.
+    :type release_request: dict(DatasetDocument) 
+    :return: A dataset document that contains the type and info of the dataset
+    :rtype: dict(DatasetDocument) 
+    """
+    # Call helper
+    released_dataset = _release_register_helper(release_request, RELEASED_DATASETS, True)
+
+    print(RELEASED_DATASETS.keys())
+
+    return released_dataset
+
 #%%
 def register(dataset):
-    dataset_name = dataset["dataset_name"]
+    """Register a new dataset
+
+    :param dataset: The dataset to read and budget to use.
+    :type dataset: dict(DatasetDocument) 
+    :return: A dataset document that contains the type and info of the dataset
+    :rtype: dict(DatasetDocument) 
+    """
 
-    if dataset_name in DATASETS:
-        abort(401, "Dataset id {} already exists. Identifies must be unique".format(dataset_name))
+    # Call helper
+    dataset = _release_register_helper(dataset, DATASETS, False)
 
-    # Add key if possible
-    if dataset["dataset_type"] not in KNOWN_DATASET_TYPE_KEYS:
-        abort(402, "Given type was {}, must be either csv_details or dataverse_details.".format(str(dataset["dataset_type"])))
+    print(DATASETS.keys())
+
+    return {"result": dataset["dataset_name"]}
+
+def _release_register_helper(dataset_request, dataset_storage, track_users):
+    # Dataset name, for convenience
+    dataset_name = dataset_request["dataset_name"]
+
+    # Check secret here, to make sure this comes from service
+
+    # Check duplicate
+    if dataset_name in dataset_storage:
+        abort(401, "Dataset id {} already exists. Identifiers must be unique".format(dataset_name))
+
+    # Check key
+    if dataset_request["dataset_type"] not in KNOWN_DATASET_TYPE_KEYS:
+        abort(402, "Given type was {}, must be either csv_details or dataverse_details.".format(str(dataset_request["dataset_type"])))
     
-    # Add budget if possible 
-    if dataset["budget"]:
-        if dataset["budget"] <= 0.0: abort(403, "Budget must be greater than 0.")
+    # Check budget 
```

Comment:
You mentioned release_budget? Where is that set?

Reply:
only set by document owner, handled in release